### PR TITLE
Bmm350

### PR DESCRIPTION
**Pull request by jlaitine**
This adds a driver for BMM350 magnetometer.

The chip is used over i2c, and polled at 20Hz frequency, similar to the bmm150. The performance settings are set to 100Hz ODR, 4 AVG. This should get good performance and relatively accurate timing.

### CODE REVIEW

File: src/drivers/magnetometer/bosch/bmm350/BMM350_registers.hpp [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+};
+
+enum PMU_CMD_AGGR_AVG : uint8_t {
+	NO_AVG = 0,                         // No average
+	AVG_2 = 1,                          // Average between 2 samples
+	AVG_4 = 2,                          // Average between 4 samples
+	AVG_8 = 3,                          // Average between 8 samples
+};
+
+union PMU_CMD_AGGR {
```
Comment (pussuw) [inline src/drivers/magnetometer/bosch/bmm350/BMM350_registers.hpp:129]: Compiler dependent behavior, are these transferred over i2c ? If so, then I would suggest not using bit fields.

Comment (jlaitine) [reply src/drivers/magnetometer/bosch/bmm350/BMM350_registers.hpp:129]: I can add __attribute__((packed))

Comment (pussuw) [reply src/drivers/magnetometer/bosch/bmm350/BMM350_registers.hpp:129]: It doesn't help here as there is no mention in the C standard about the storage of bitfields, their order can be chosen by the compiler at will.

Comment (jlaitine) [reply src/drivers/magnetometer/bosch/bmm350/BMM350_registers.hpp]: Right, cool. Removed bitfields now.

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp [hunk 2]
```diff
@@ -0,0 +1,687 @@
...
+	float cr_ax_comp_x = (xyzt[0] - mag_comp.cross_axis.cross_x_y * xyzt[1]) /
+			     (1 - mag_comp.cross_axis.cross_y_x * mag_comp.cross_axis.cross_x_y);
+	float cr_ax_comp_y = (xyzt[1] - mag_comp.cross_axis.cross_y_x * xyzt[0]) /
+			     (1 - mag_comp.cross_axis.cross_y_x * mag_comp.cross_axis.cross_x_y);
+	float cr_ax_comp_z = (xyzt[2] +
+			      (xyzt[0] * (mag_comp.cross_axis.cross_y_x * mag_comp.cross_axis.cross_z_y -
+					  mag_comp.cross_axis.cross_z_x) - xyzt[1] *
+			       (mag_comp.cross_axis.cross_z_y - mag_comp.cross_axis.cross_x_y *
+				mag_comp.cross_axis.cross_z_x)) /
+			      (1 - mag_comp.cross_axis.cross_y_x * mag_comp.cross_axis.cross_x_y));
```
Comment (pussuw) [inline src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:256]: Can these result in division by zero?

**Review by pussuw — APPROVED** [review]
One question, otherwise LGTM!

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp (hunk 2 above)
Comment (jlaitine) [reply src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:256]: I don't think so. The compensation values are very small floats, and the dividers are always close to 1. The equations are copied 1:1 from bosch reference

**Review by jpaali — APPROVED** [review]

File: src/drivers/magnetometer/bosch/bmm350/BMM350_registers.hpp [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+#include <cstdint>
+
+namespace Bosch_BMM350
+{
+static constexpr uint32_t I2C_SPEED          = 400000;
+static constexpr uint8_t I2C_ADDRESS_DEFAULT = 0x14;
+
+static constexpr uint8_t DEFAULT_ID          = 0x33;
+
+static constexpr int16_t OVERFLOW_XYAXES     = -4096;
```
Comment (jlaitine) [inline src/drivers/magnetometer/bosch/bmm350/BMM350_registers.hpp:52]: These were actually some lines which I copy paster accidentally from some other driver. They are meaningless, and should be removed. Will make a small fix for that still.

**Pull request merged**